### PR TITLE
BugFix: #42068: Support Field Removals for IstioOperator

### PR DESCRIPTION
**Pull request by Kaushal28**
**Please provide a description of this PR:**

Issue https://github.com/istio/istio/issues/42068

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [X] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by Kaushal28** [comment]
I figured out the root cause. It is not using server side apply (SSA) currently. In client side apply, it is using `github.com/evanphx/json-patch/v5` library, that is implementing JSON patch merge based on this RFC: https://www.rfc-editor.org/rfc/rfc7396. According to [this](https://www.rfc-editor.org/rfc/rfc7396#:~:text=Null%20values%20in%20the%20merge%0A%20%20%20patch%20are%20given%20special%20meaning%20to%20indicate%20the%20removal%20of%20existing%0A%20%20%20values%20in%20the%20target.), JSON patch needs to set `null` in the value of the key that needs to be removed from the target JSON object. However, in this case, user is completely removing the field and hence that field is not getting updated in the target JSON.

One option is to use server side apply, which is available from K8s 1.16+. As the latest versions of istio only support 1.16+ K8s version ([reference](https://istio.io/latest/docs/releases/supported-releases/#:~:text=1.23%2C%201.24%2C%201.25%2C%201.26)), I believe this PR will fix the issue for them. For the earlier versions, a fix for this will increase client side complexity as we will have to somehow figure out a way to determine removed fields and set `null` for them.

I have tested this change in my local, I can also add a few tests for this if this change makes sense. Let me know your thoughts.

File: operator/pkg/helmreconciler/common.go [hunk 1]
```diff
@@ -116,33 +114,6 @@ func buildInstallTreeString(componentName name.ComponentName, prefix string, sb
 	}
 }
 
-// applyOverlay applies an overlay using JSON patch strategy over the current Object in place.
-func applyOverlay(current, overlay *unstructured.Unstructured) error {
```
Comment (howardjohn) [inline operator/pkg/helmreconciler/common.go:120]: I am convinced for the installed-state IstioOperator we should remove this. I am not so sure it is correct for *all* purposes. especially if we are doing SSA or this for other types.

Should we do this only for the IstioOperator?

Comment (Kaushal28) [reply operator/pkg/helmreconciler/common.go]: Makes sense. I have added a condition to only do this for `IstioOperator` and also added test cases to test field removals.

**Comment by howardjohn** [comment]
LGTM, assigned some folks more familiar with this though

**Review by zirain — APPROVED** [review]
lgtm, add a hold for @richardwxn @ostromart

**Comment by Kaushal28** [comment]
@richardwxn @ostromart Could you please review the changes? Any suggestions/feedback would be really helpful. Thanks.

**Review by hanxiaop — COMMENTED** [review]
Can you add a release note?

**Comment by Kaushal28** [comment]
@hanxiaop added a release note.

**Review by hanxiaop — APPROVED** [review]

**Comment by Kaushal28** [comment]
@howardjohn @richardwxn @ostromart Could you please review the changes? Any suggestions/feedback would be really helpful. Thanks.

**Review by richardwxn — APPROVED** [review]

**Review by cnvergence — APPROVED** [review]
rechecked the issue, I can confirm that now it works as expected
Thanks @Kaushal28

**Comment by Kaushal28** [comment]
@cnvergence @zirain is it now mergeable then? Let me know any thing is still pending.

**Comment by Kaushal28** [comment]
I ran `make gen` after all the approvals and it changed comments in a few generated files and added a new reviewer. Could anyone please have a look?

**Comment by Kaushal28** [comment]
Hi @zirain, I ran `make gen` after all the approvals and it changed comments in a few generated files and added a new reviewer. Could you please have a look?

**Review by morvencao — APPROVED** [review]

**Comment by Kaushal28** [comment]
Can anyone please review and provide feedback on this PR?

**Comment by Kaushal28** [comment]
Hi @howardjohn, @cnvergence Could you please look into this PR? Merge is pending since long time just because of changes in comments of a few auto generated files. Let me know if anything else is missing. Thanks.

**Comment by cnvergence** [comment]
@Kaushal28 please review the conflicts

**Pull request merged**